helix-view/src/gutter.rs

```diff
@@ -91,24 +126,41 @@ pub fn line_numbers<'doc>(
             } else {
                 line + 1
             };
+
+            let n_digits = count_digits(display_num);
+
             let style = if selected && is_focused {
                 linenr_select
             } else {
                 linenr
             };
-            write!(out, "{:>1$}", display_num, width).unwrap();
+
+            // if line number overflows maximum alotted width, truncate start
+            if n_digits > width {
+                let display_trailing = (display_num as u32) % 10_u32.pow((width - 1) as u32);
+                write!(out, "{}{:0>2$}", ELLIPSIS, display_trailing, width - 1).unwrap();
+            } else {
+                write!(out, "{:>1$}", display_num, width).unwrap();
+            }
```

Comment:
I don't think we need truncation, line numbers should always be fully visible. I can't think of any other editor truncating, past 8 digits the file will be enormously large anyway.

Reply:
Alright, that's easy enough. The design was taken from the discussion in #2583 - happy to change it if it's preferred that it never truncates.